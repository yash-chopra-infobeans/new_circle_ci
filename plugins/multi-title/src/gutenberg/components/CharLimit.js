export const CharLimit = ({ showCharLimitAt = 0, charLimit, charLimitValue = '' }) => {
  // if no character limit just return
  if (!charLimit) {
    return null;
  }

  /**
   * calculate width the coloured line should be based on value length
   * as a percentage of character limit.
   */
  let width = `${(charLimitValue.length / charLimit) * 100}%`;

  let showChar = showCharLimitAt;
  if (showChar < 0 && charLimit) {
    showChar = charLimit / 2;
  }

  let backgroundColor = '#6c7781';
  // to highlight remaining characters if less than half of the limit
  if (charLimitValue.length >= charLimit / 2) {
    backgroundColor = '#FFB900';
  }

  // warning char limit hit
  if (charLimitValue.length === charLimit) {
    backgroundColor = '#F56E28';
  }

  // potential error beyond char limit
  if (charLimitValue.length > charLimit) {
    backgroundColor = '#DC3232';
    // setting with to 100% so the bar doesn't go beyond the div
    width = '100%';
  }

  return (
    <div className="wrappedInput-charLimit charLimit">
      <div className="charLimit-count" style={{ backgroundColor }}>
        {`${charLimit - charLimitValue.length}`}
      </div>
      <div className="charLimit-bar">
        <div className="charLimit-bar-progress" style={{ backgroundColor, width }} />
      </div>
    </div>
  );
};

export default CharLimit;
