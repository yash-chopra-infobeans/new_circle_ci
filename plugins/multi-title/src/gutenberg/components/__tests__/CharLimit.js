import { shallow } from 'enzyme';
import CharLimit from '../CharLimit';

describe('<CharLimit />', () => {
  it('should render component without errors when only passing charLimit prop', () => {
    const componentProps = {
      charLimit: 10,
    };
    const component = shallow(<CharLimit {...componentProps} />);
    const bar = component.find('.charLimit-count');

    expect(component).toMatchSnapshot();
    expect(bar.text()).toEqual('10');
    expect(bar.prop('style')).toEqual({ backgroundColor: '#6c7781' });
  });

  it('should have a grey bar when value is less than half of the character limit', () => {
    const componentProps = {
      showCharLimitAt: -1,
      charLimit: 10,
      charLimitValue: 'test',
    };
    const component = shallow(<CharLimit {...componentProps} />);
    const bar = component.find('.charLimit-count');

    expect(component).toMatchSnapshot();
    expect(bar.text()).toEqual('6'); // 10 (charLimit) -4 (charLimitValue length) = 6
    expect(bar.prop('style')).toEqual({ backgroundColor: '#6c7781' });
  });

  it('should have an orange bar when the character limit is reached', () => {
    const componentProps = {
      showCharLimitAt: -1,
      charLimit: 10,
      charLimitValue: '0123456789', // 10 characters to match charLimit
    };
    const component = shallow(<CharLimit {...componentProps} />);
    const bar = component.find('.charLimit-count');

    expect(component).toMatchSnapshot();
    expect(bar.text()).toEqual('0');
    expect(bar.prop('style')).toEqual({ backgroundColor: '#F56E28' });
  });

  it('should have a red bar when the character limit has been exceeded', () => {
    const componentProps = {
      showCharLimitAt: -1,
      charLimit: 10,
      charLimitValue: 'this is more than ten characters',
    };
    const component = shallow(<CharLimit {...componentProps} />);
    const bar = component.find('.charLimit-count');

    expect(component).toMatchSnapshot();
    expect(bar.text()).toEqual('-22'); // 10 (charLimit) -32 (charLimitValue length) = -22
    expect(bar.prop('style')).toEqual({ backgroundColor: '#DC3232' });
  });
});
