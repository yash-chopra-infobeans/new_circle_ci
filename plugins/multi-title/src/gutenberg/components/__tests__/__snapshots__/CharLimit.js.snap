// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharLimit /> should have a grey bar when value is less than half of the character limit 1`] = `
<div
  className="wrappedInput-charLimit charLimit"
>
  <div
    className="charLimit-count"
    style={
      Object {
        "backgroundColor": "#6c7781",
      }
    }
  >
    6
  </div>
  <div
    className="charLimit-bar"
  >
    <div
      className="charLimit-bar-progress"
      style={
        Object {
          "backgroundColor": "#6c7781",
          "width": "40%",
        }
      }
    />
  </div>
</div>
`;

exports[`<CharLimit /> should have a red bar when the character limit has been exceeded 1`] = `
<div
  className="wrappedInput-charLimit charLimit"
>
  <div
    className="charLimit-count"
    style={
      Object {
        "backgroundColor": "#DC3232",
      }
    }
  >
    -22
  </div>
  <div
    className="charLimit-bar"
  >
    <div
      className="charLimit-bar-progress"
      style={
        Object {
          "backgroundColor": "#DC3232",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;

exports[`<CharLimit /> should have an orange bar when the character limit is reached 1`] = `
<div
  className="wrappedInput-charLimit charLimit"
>
  <div
    className="charLimit-count"
    style={
      Object {
        "backgroundColor": "#F56E28",
      }
    }
  >
    0
  </div>
  <div
    className="charLimit-bar"
  >
    <div
      className="charLimit-bar-progress"
      style={
        Object {
          "backgroundColor": "#F56E28",
          "width": "100%",
        }
      }
    />
  </div>
</div>
`;

exports[`<CharLimit /> should render component without errors when only passing charLimit prop 1`] = `
<div
  className="wrappedInput-charLimit charLimit"
>
  <div
    className="charLimit-count"
    style={
      Object {
        "backgroundColor": "#6c7781",
      }
    }
  >
    10
  </div>
  <div
    className="charLimit-bar"
  >
    <div
      className="charLimit-bar-progress"
      style={
        Object {
          "backgroundColor": "#6c7781",
          "width": "0%",
        }
      }
    />
  </div>
</div>
`;
